### PR TITLE
Middlewares auth

### PR DESCRIPTION
**Pull request by mkkasem**
I created the middlewares but did not add them to the routes files yet

### CODE REVIEW

File: src/middlewares/item-authorization.js
Comment (AmmarBaki2) [inline src/middlewares/item-authorization.js:7]: Since the `item.owner` is the reference to the user you don't need to use `._id`, 
`item.owner.toString()` is enough.

File: src/middlewares/user-authorization.js
Comment (AmmarBaki2) [inline src/middlewares/user-authorization.js:7]: Good job 👏 
We can use this when a user wants to delete his account

**Review by AmmarBaki2 — CHANGES_REQUESTED** [review]
Excellent job @mkkasem 👏 
There is a slight change I requested in the comments below

**Review by AmmarBaki2 — APPROVED** [review]
Well done 😀

**Review by mehmetfatiherdem — APPROVED** [review]

**Comment by mkkasem** [comment]
@Shrreya  can you please have a look at my code when you are free 😁?

**Comment by Shrreya** [comment]
> @Shrreya can you please have a look at my code when you are free 😁?

@mkkasem Are all the changes done? I had started reviewing but then I saw you were pushing few commits so I thought I'll wait.

**Comment by mkkasem** [comment]
> > @Shrreya can you please have a look at my code when you are free 😁?
> 
> @mkkasem Are all the changes done? I had started reviewing but then I saw you were pushing few commits so I thought I'll wait.

Yes I think I am done, but I haven't add these functions to the routes because I am not sure if I should add them now or the next sprint. I was planning to ask you about this issue after you approve this PR.

**Review by Peri7at — APPROVED** [review]

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:4]: Better to rename this to: `validTokenWithValidTime`

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:12]: Both these tokens are the same. Properties of one need to be changed.
Also rename one to `validTokenWithInvalidTime` and other to `invalidTokenWithValidTime`.

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:21]: You can rename this to `mockUser` which explains that this will be used as mock value in tests.

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:36]: You can rephrase this to `Should authenticate for valid token with valid expiry time`.

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:50]: You should also check that `next()` has been called and `req.user` value is set.

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:54]: You can rephrase this to `Should not authenticate for valid token with invalid expiry time`

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/isAuth.test.js:28]: You should add `mockReturnThis()` only for functions that will be chained in your implementation.

File: src/middlewares/__tests__/isAuth.test.js
Comment (Shrreya) [reply src/middlewares/__tests__/isAuth.test.js:54]: Check if other test statements also need to rephrased. Try to follow similar words throughout test statements.

File: src/middlewares/__tests__/item-authorization.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/item-authorization.test.js:30]: Try to avoid mentioning implementation logic in test statements.
So here you can rephrase to `Should authorize when requesting user is the owner of requested item`.

Comment (Shrreya) [reply src/middlewares/__tests__/item-authorization.test.js:30]: Check the other test statements too and see if there's any rephrasing required similar to this.

**Review by Shrreya — CHANGES_REQUESTED** [review]
@mkkasem Good job with the implementation! Logic looks solid.

Some improvements related to best practices are required.

**Comment by Shrreya** [comment]
@mkkasem Can you also merge the main branch into your branch? To ensure latest changes from your team mates are included.

File: src/middlewares/isAuth.js
Comment (mkkasem) [reply src/middlewares/isAuth.js:6]: yes, I was getting the user without deconstruction, but it was returning undefined because the verify function returns an object that contains the user object along with some properties like expiration time that's why I used the deconstruction

File: src/middlewares/isAuth.js
Comment (Shrreya) [reply src/middlewares/isAuth.js:9]: I think you can fix that by removing `async` in front of the arrow function, since there's no `await` call inside.

Comment (mkkasem) [reply src/middlewares/isAuth.js:9]: I did that but still getting the same error.

File: src/middlewares/user-authorization.js
Comment (mkkasem) [reply src/middlewares/user-authorization.js:13]: Put  /api/users/:id
a user can only update his information 
also for deletion

File: package.json
Comment (AmmarBaki2) [inline package.json]: @mkkasem if you look at the discussion between me and Shrreya on the #23 PR you will get a better view on this. 
`set` doesn't work in github, you need to use it locally but when you commit make sure that it is `export` not `set` 
Just when you are going to push them to github after committing you can revert it back to `set` so it works on you terminal

Comment (AmmarBaki2) [inline package.json]: @mkkasem if you look at the discussion between me and Shrreya on the https://github.com/RCDD-202110-TUR-BEW/backend-capstone-turkey-paying-it-forward/pull/23 PR you will get a better view on this.
set doesn't work in github, you need to use it locally but when you commit make sure that it is export not set
Just when you are going to push them to github after committing you can revert it back to set so it works on you terminal

**Review by AmmarBaki2 — CHANGES_REQUESTED** [review]

File: package.json
Comment (mkkasem) [reply package.json]: I saw the conversation and I know that I need to set it back to export, which I did in the last commit because I wasn't able to commit , so I fixed it on github immediately but I didn't know that if I change it to export without commiting it could work. 
Thanks ammar

Comment (AmmarBaki2) [reply package.json]: You are welcome 😊 
I faced the same problem at first, then I noticed that when you are pushing (as you know it runs the test script), it uses the script that is present in your local repo despite what is present in the commit.

File: src/middlewares/isAuth.js
Comment (mehmetfatiherdem) [reply src/middlewares/isAuth.js]: I got the same error in another function. I solved it by adding a res.json()

File: src/middlewares/__tests__/user-authentication.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/user-authentication.test.js:59]: Here as well try to keep the order of execution in the order of your expect statements.

So first check `req.user`, then do all the checks on `verify` and last on `next` and `res.json` or `res.status`.

Also always run `toHaveBeenCalled` first, then `toHaveBeenCalledTimes`, then `toHaveBeenCalledWith` and last `toHaveReturnedWith` or `toThrow`.

File: src/middlewares/__tests__/user-authorization.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:20]: Let's keep our language gender-neutral. So you can say `Should authorize user when trying to modify own information`.

File: src/middlewares/__tests__/user-authorization.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:55]: Better statement could be `Should unauthorize user when user ID is not found`

File: src/middlewares/__tests__/user-authorization.test.js
Comment (Shrreya) [inline src/middlewares/__tests__/user-authorization.test.js:67]: Seems like `to` added twice by mistake.

File: src/middlewares/item-authorization.js
Comment (Shrreya) [inline src/middlewares/item-authorization.js:2]: You can shorten this as:
```
/* this middleware is called after the user-authentication middleware to verify
 that user is authorized to access some protected item endpoints */

File: src/middlewares/user-authentication.js
Comment (Shrreya) [inline src/middlewares/user-authentication.js:1]: It would be better to write this as:
```
/* this middleware is called on all routes except the global ones to verify that user is authenticated */

File: src/middlewares/user-authorization.js
Comment (Shrreya) [inline src/middlewares/user-authorization.js:2]: Rewrite this similar to the comment above about item-authorization.

**Comment by mkkasem** [comment]
@Shrreya all changes are done and I merged the main branch also.
I hope it's ready to be merged now :)

**Comment by mkkasem** [comment]
@Ammar-64 Could you please merge this branch when you are free?

**Review by AmmarBaki2 — APPROVED** [review]

**Pull request merged**
